core/src/main/scala/org/apache/spark/TaskContextImpl.scala

```diff
@@ -57,57 +68,75 @@ private[spark] class TaskContextImpl(
   // Whether the task has failed.
   @volatile private var failed: Boolean = false
```

Comment:
This need not be volatile anymore - given that it is updated and queried within a synchronized block.
We could revisit for completed too - though that would be an extension.